clients/core/src/main/scala/mqperf/Sender.scala

```diff
@@ -24,44 +25,53 @@ class Sender(config: Config, mq: Mq, clock: Clock) extends StrictLogging {
 
     Future {
       blocking {
-        val start = clock.millis()
-        val end = start + config.testLengthSeconds * 1000L
-
-        val permits = new AtomicInteger(config.maxSendInFlight)
-        val batchSize = config.batchSizeSend
-
+        val end = clock.millis() + config.testLengthSeconds.seconds.toMillis
         while (clock.millis() < end) {
           val iterationStart = clock.millis()
-          // we send at most `config.msgsPerSecond` messages, but no more than the number of available permits
-          val msgsToSend = math.min(config.msgsPerSecond, permits.getAndUpdate(p => math.max(0, p - config.msgsPerSecond)))
-          logger.info(s"Messages to send: $msgsToSend")
-          val batches = msgsToSend / batchSize
-          (1 to batches).foreach { _ =>
-            val sendStart = clock.millis()
-            mqSender.send(nextMessagesBatch()).onComplete { result =>
-              permits.addAndGet(batchSize)
-              LocalMetrics.messageCounter.inc(batchSize.toDouble)
-              LocalMetrics.messageLatencyHistogram.observe((clock.millis() - sendStart).toDouble)
-              result match {
-                case Failure(t) => logger.error("Exception when sending a batch of messages", t)
-                case Success(_) =>
-              }
-            }
-          }
+          logger.info(s"Messages to send: ${config.msgsPerSecond} (concurrency=${config.senderConcurrency})")
+          runIterations(mqSender)
```

Comment:
we never wait for the future to finish, so this will just start a new `runIterations` every second (previously we were blocking on `permits.getAndUpdate`)

so I think here we would need to replace this with `.flatMap`s and recursive calls, similarly as in the sender

Reply:
alternatively, we can replace this with a bunch of sender processes which send messages when permits are available (from a semaphore), plus a separate process which replenishes these permits every second; but to do this in a non-blocking way, we'd need sth like cats-effect or ZIO (which is definitely viable)

Reply:
I added loop which increases number of permits every second. It runs in dedicated single-thread ExecutionContext. If it is not enough, I can improve it by using cats.